File: C++/Patterns/MVC/common/Controller.cpp
#include "MVC.hpp"

namespace oobench {

    Controller::Controller(View& theView): view(&theView) {
        view->getModel().attach(*this);
    }

    Controller::Controller(const Controller& anotherController) {
        view = anotherController.view;
        view->getModel().attach(*this);
    }

    void Controller::handleEvent(ControllerEvent& theEvent) {
    }

    void Controller::update() {
    }

    Controller::~Controller() {
        view->getModel().detach(*this);
    }
}

File: Java/org/oobench/patterns/mvc/common/Controller.java
/*
 * Base class for controllers.
 */

package org.oobench.patterns.mvc.common;

public abstract class Controller extends Observer {

    private View view;

    public Controller(View theView) {
        view = theView;
        view.getModel().attach(this);
    }

    public abstract void handleEvent(ControllerEvent theEvent);

    public void unregister() {
        view.getModel().detach(this);
    }

    public View getView() {
        return view;
    }
}
